ros_controllers/src/diff_drive_controller.cpp

```diff
@@ -14,12 +14,272 @@
+            return CallbackReturn::FAILURE;
+         }
+      }
+
+      return CallbackReturn::SUCCESS;
+   };
+
+   if (auto robot_hardware = robot_hardware_.lock()) {
+      const auto left_result =
+         configure_side("left", left_wheel_names_, registered_left_wheel_handles_, robot_hardware);
+      const auto right_result =
+         configure_side("right", right_wheel_names_, registered_right_wheel_handles_, robot_hardware);
+
+      if (left_result == CallbackReturn ::FAILURE or right_result == CallbackReturn::FAILURE) {
+         return CallbackReturn::FAILURE;
+      }
+
+      registered_operation_mode_handles_.resize(write_op_names_.size());
+      for (size_t index = 0; index < write_op_names_.size(); ++index) {
+         const auto op_name = write_op_names_[index].c_str();
+         auto& op_handle = registered_operation_mode_handles_[index];
+
+         auto result = robot_hardware->get_operation_mode_handle(op_name, &op_handle);
+         if (result != hardware_interface::HW_RET_OK) {
+            RCLCPP_WARN(logger, "unable to obtain operation mode handle for %s", op_name);
+            return CallbackReturn::FAILURE;
+         }
+      }
+
+   } else {
+      return CallbackReturn::ERROR;
+   }
+
+   if (registered_left_wheel_handles_.empty() or registered_right_wheel_handles_.empty() or
+       registered_operation_mode_handles_.empty()) {
+      return CallbackReturn::ERROR;
+   }
+
+   wheel_params_.wheels_per_side = registered_left_wheel_handles_.size();
+
+   left_previous_commands_ = std::vector<double>(0, left_wheel_names_.size());
+   right_previous_commands_ = std::vector<double>(0, right_wheel_names_.size());
+
+   velocity_msg_ptr_ = std::make_shared<Twist>();
+
+   auto callback = [this](const std::shared_ptr<Twist> msg) -> void {
+      if (subscriber_is_active_) {
+         *velocity_msg_ptr_ = *msg;
+      }
+   };
+
+   velocity_command_subscriber_ =
+      lifecycle_node_->create_subscription<Twist>(DEFAULT_COMMAND_TOPIC, rclcpp::SystemDefaultsQoS(), callback);
+
+   set_op_mode(hardware_interface::OperationMode::INACTIVE);
+
+   return CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+DiffDriveController::on_activate(const rclcpp_lifecycle::State&)
+{
+   is_halted = false;
+   subscriber_is_active_ = true;
+   //   traj_point_active_ptr_ = &traj_external_point_ptr_;
+   return CallbackReturn::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+DiffDriveController::on_deactivate(const rclcpp_lifecycle::State&)
 {
-  constexpr const char kDefaultCmdVelTopic[] = "cmd_vel";
+   subscriber_is_active_ = false;
+   return CallbackReturn::SUCCESS;
 }
 
-namespace ros_controllers
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+DiffDriveController::on_cleanup(const rclcpp_lifecycle::State&)
 {
+   velocity_msg_ptr_ = std::make_shared<Twist>();
```

Comment:
use [`reset()`](https://en.cppreference.com/w/cpp/memory/shared_ptr/reset)

Reply:
Reset would leave the pointer in a nullptr state. I think this call is meant to set the values to their initial states, so I thought this was a convenient (albeit inefficient) way to do that. I could reassign a a freshly constructed vector3 to both angular and linear, or just manually assign 0's to all of the members. What do you think?

Reply:
I see, if that's the case, this is fine